util/kubeconfig/kubeconfig.go

```diff
@@ -174,3 +144,98 @@ func GenerateSecretWithOwner(clusterName client.ObjectKey, data []byte, owner me
 		},
 	}
 }
+
+// NeedsClientCertRotation returns whether any of the Kubeconfig secret's client certificates will expire before the given threshold.
+func NeedsClientCertRotation(configSecret *corev1.Secret, threshold time.Duration) (bool, error) {
```

Comment:
I wonder if this check is necessary during kubeconfig reconciliation in the cluster controller: https://github.com/kubernetes-sigs/cluster-api/blob/c437644da00406b515201a48b5b93f29b09d3878/controllers/cluster_controller_phases.go#L229

Reply:
Same issue being discussed in the main thread here https://github.com/kubernetes-sigs/cluster-api/pull/3140#issuecomment-639528788